nanofi/src/core/string_utils.c

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/cstructs.h"
+#include "core/string_utils.h"
+#include <string.h>
+#include <stdlib.h>
+
+int validate_list(struct token_list * tk_list) {
+    if (tk_list && tk_list->head && tk_list->tail && tk_list->size > 0) {
+        return 1;
+    }
+    return 0;
+}
+
+void add_token_to_list(struct token_list * tk_list, char * begin, uint64_t len) {
+    struct token_node * new_node = (struct token_node *)malloc(sizeof(struct token_node));
+    new_node->data = (char *)malloc((len+1) * sizeof(char));
+    strncpy(new_node->data, begin, len);
+    new_node->data[len] = '\0';
+    new_node->next = NULL;
+
+    if (!tk_list->head) {
+        tk_list->head = tk_list->tail = new_node;
+        tk_list->size = 1;
+        tk_list->total_bytes = (len);
+        return;
+    }
+
+    tk_list->tail->next = new_node;
+    tk_list->tail = new_node;
+    tk_list->size++;
+    tk_list->total_bytes += (len);
+}
+
+void free_token_node(struct token_node * node) {
+    if (node) {
+        free(node->data);
+    }
+    free(node);
+    memset(node, 0, sizeof(struct token_node));
+}
+
+void free_all_tokens(struct token_list * tks) {
+    while (tks && tks->head) {
+        struct token_node * node = tks->head;
+        tks->head = tks->head->next;
+        free_token_node(node);
+    }
+}
+
+void remove_last_node(token_list * tks) {
+    if (!validate_list(tks)) {
+        return;
+    }
+
+    if (tks->size == 1 || tks->head == tks->tail) {
+        free_all_tokens(tks);
+        memset(tks, 0, sizeof(struct token_list));
+        return;
+    }
+
+    struct token_node * tmp_head = tks->head;
+    struct token_node * tmp_tail = tks->tail;
+
+    while (tmp_head->next && (tmp_head->next != tmp_tail)) {
+        tmp_head = tmp_head->next;
+    }
+
+    struct token_node * tail_node = tmp_tail;
+    tks->tail = tmp_head;
+    tks->tail->next = NULL;
+
+    tks->size--;
+    tks->total_bytes -= (strlen(tail_node->data));
+    free_token_node(tail_node);
+}
+
+void attach_lists(token_list * to, token_list * from) {
+    if (to && validate_list(from)) {
+        if (!to->head) {
+            to->head = from->head;
+            to->tail = from->tail;
+            to->size = from->size;
+            to->total_bytes = from->total_bytes;
+            return;
+        }
+
+        if (!to->tail) return;
+
+        to->tail->next = from->head;
+        to->tail = from->tail;
+        to->size += from->size;
+        to->total_bytes += from->total_bytes;
+    }
+}
+
+token_list tokenize_string(const char * str, char delim) {
+    token_list tks;
+    memset(&tks, 0, sizeof(struct token_list));
+
+    if (!str) return tks;
+
+    char * begin = (char *)str;
```

Comment:
Please leave this const!
Pointer to const array (aka const char *) can be incremented as you need. 
Const pointers to char * (akar char * const) cannot, but you don't have any of those.